File: crush-algorithm/src/crush.c
#include "crush.h"

// 初始化集群
void init_cluster(Cluster* cluster) {
    cluster->node_count = 0;
}

// 向集群中添加节点
void add_node(Cluster* cluster, int id, const char* name) {
    if (cluster->node_count < MAX_NODES) {
        cluster->nodes[cluster->node_count].id = id;
        strncpy(cluster->nodes[cluster->node_count].name, name, sizeof(cluster->nodes[cluster->node_count].name) - 1);
        cluster->node_count++;
    }
}

// 简单的哈希函数，用于根据对象名称进行映射
int hash(const char* str) {
    int hash = 0;
    while (*str) {
        hash = (hash * 31) + *str++;
    }
    return hash;
}

// 使用 CRUSH 算法映射数据对象到集群节点
int crush_map(Cluster* cluster, const char* object_name) {
    int hashed_value = hash(object_name);
    return hashed_value % cluster->node_count;
}

File: crush-algorithm-go/src/crush.go
package main

import (
	"fmt"
	"hash/fnv"
	"strings"
)

const MaxNodes = 1000

// Node 结构体代表一个集群节点
type Node struct {
	ID   int
	Name string
}

// Cluster 结构体表示整个集群
type Cluster struct {
	Nodes     []Node
	NodeCount int
}

// 初始化集群
func InitCluster() *Cluster {
	return &Cluster{
		Nodes:     make([]Node, 0, MaxNodes),
		NodeCount: 0,
	}
}

// 向集群添加节点
func (cluster *Cluster) AddNode(id int, name string) {
	if cluster.NodeCount < MaxNodes {
		node := Node{ID: id, Name: name}
		cluster.Nodes = append(cluster.Nodes, node)
		cluster.NodeCount++
	}
}

// 简单的哈希函数（FNV-1a）
func hash(str string) uint32 {
	h := fnv.New32a()
	h.Write([]byte(str))
	return h.Sum32()
}

// CRUSH 映射函数，将数据映射到集群节点
func (cluster *Cluster) CrushMap(objectName string) int {
	hashedValue := hash(objectName)
	return int(hashedValue) % cluster.NodeCount
}
